sql/expression/function/math.go

```diff
@@ -113,282 +114,244 @@ func (r *Rand) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+		val, err := sql.Uint64.Convert(arg)
 
+		if err != nil {
+			return nil, err
+		}
 
-// TanFuncLogic is a UnaryMathFloatFuncWrapper implementation which returns the tan of a value
-type TanFuncLogic struct{}
+		bytes = []byte(strconv.FormatUint(val.(uint64), 10))
+	case float32:
+		s := floatToString(float64(val))
+		bytes = []byte(s)
+	case float64:
+		s := floatToString(val)
+		bytes = []byte(s)
+	case bool:
+		if val {
+			bytes = []byte{1}
+		} else {
+			bytes = []byte{0}
+		}
+	default:
+		return nil, ErrInvalidArgument.New("crc32", fmt.Sprint(arg))
+	}
 
-// EvalFloat handles the case where the number argument passed is a float
-func (fl TanFuncLogic) EvalFloat(n float64) (interface{}, error) {
-	return math.Tan(n), nil
+	return crc32.ChecksumIEEE(bytes), nil
 }
 
+func hexChar(b byte) byte {
+	 if b > 9 {
+		return b - 10 + byte('A')
+	 }
 
-// ASinFuncLogic is a UnaryMathFloatFuncWrapper implementation which returns the asin of a value
-type ASinFuncLogic struct{}
-
-// EvalFloat handles the case where the number argument passed is a float
-func (fl ASinFuncLogic) EvalFloat(n float64) (interface{}, error) {
-	return math.Asin(n), nil
+	 return b + byte('0')
 }
 
+func hexForFloat(f float64) (string, error) {
+	if f < 0 {
+		f -= 0.5
+		n := int64(f)
+		bytes := *(*[8]byte)(unsafe.Pointer(&n))
+		for i := 0; i < 4; i++ {
+			bytes[i], bytes[7-i] = bytes[7-i], bytes[i]
+		}
 
-// ACosFuncLogic is a UnaryMathFloatFuncWrapper implementation which returns the acos of a value
-type ACosFuncLogic struct{}
+		return fmt.Sprintf("%X", bytes), nil
+	}
 
-// EvalFloat handles the case where the number argument passed is a float
-func (fl ACosFuncLogic) EvalFloat(n float64) (interface{}, error) {
-	return math.Acos(n), nil
+	f += 0.5
+	n := uint64(f)
+	return fmt.Sprintf("%X", n), nil
 }
 
+func HexFunc(ctx *sql.Context, arg interface{}) (interface{}, error) {
+	switch val := arg.(type) {
+	case string:
+		buf := make([]byte, 0, 2*len(val))
+		for _, c := range val {
+			high := byte(c / 16)
+			low := byte(c % 16)
 
-// ATanFuncLogic is a UnaryMathFloatFuncWrapper implementation which returns the atan of a value
-type ATanFuncLogic struct{}
+			buf = append(buf, hexChar(high))
+			buf = append(buf, hexChar(low))
+		}
 
-// EvalFloat handles the case where the number argument passed is a float
-func (fl ATanFuncLogic) EvalFloat(n float64) (interface{}, error) {
-	return math.Atan(n), nil
-}
+		return string(buf), nil
 
+	case uint8, uint16, uint32, uint, int, int8, int16, int32, int64:
+		n, err := sql.Int64.Convert(arg)
 
-// CotFuncLogic is a UnaryMathFloatFuncWrapper implementation which returns the cot of a value
-type CotFuncLogic struct{}
+		if err != nil {
+			return nil, err
+		}
 
-// EvalFloat handles the case where the number argument passed is a float
-func (fl CotFuncLogic) EvalFloat(n float64) (interface{}, error) {
-	return 1.0 / math.Tan(n), nil
-}
+		a := n.(int64)
+		if a < 0 {
+			bytes := *(*[8]byte)(unsafe.Pointer(&a))
```

Comment:
Might add a link to some docs describing this algorithm. It's kinda magic

Reply:
refactored and commented.